digdag-core/src/main/java/io/digdag/core/database/DatabaseTaskQueueServer.java

```diff
@@ -0,0 +1,495 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.time.Instant;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import com.google.common.base.*;
+import com.google.common.collect.*;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.digdag.core.queue.QueueSettingStore;
+import io.digdag.core.queue.QueueSettingStoreManager;
+import io.digdag.core.queue.StoredQueueSetting;
+import io.digdag.core.queue.ImmutableStoredQueueSetting;
+import org.skife.jdbi.v2.DBI;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.sqlobject.SqlQuery;
+import org.skife.jdbi.v2.sqlobject.SqlUpdate;
+import org.skife.jdbi.v2.sqlobject.Bind;
+import org.skife.jdbi.v2.sqlobject.GetGeneratedKeys;
+import org.skife.jdbi.v2.StatementContext;
+import org.skife.jdbi.v2.sqlobject.customizers.Mapper;
+import org.skife.jdbi.v2.tweak.ResultSetMapper;
+import io.digdag.client.config.Config;
+import io.digdag.spi.TaskRequest;
+import io.digdag.spi.TaskQueueServer;
+import io.digdag.spi.TaskStateException;
+import io.digdag.core.repository.ResourceConflictException;
+import io.digdag.core.repository.ResourceNotFoundException;
+import com.google.common.annotations.VisibleForTesting;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class DatabaseTaskQueueServer
+        extends BasicDatabaseStoreManager<DatabaseTaskQueueServer.Dao>
+        implements TaskQueueServer
+{
+    private final DatabaseTaskQueueConfig queueConfig;
+    private final ObjectMapper taskObjectMapper;
+
+    private final int expireLockInterval;
+    private final LocalLockMap localLockMap = new LocalLockMap();
+    private final ScheduledExecutorService expireExecutor;
+
+    @Inject
+    public DatabaseTaskQueueServer(DBI dbi, DatabaseConfig config, DatabaseTaskQueueConfig queueConfig, ObjectMapper taskObjectMapper)
+    {
+        super(config.getType(), Dao.class, dbi);
+        this.queueConfig = queueConfig;
+        this.taskObjectMapper = taskObjectMapper;
+        this.expireLockInterval = config.getExpireLockInterval();
+        this.expireExecutor = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("lock-expire-%d")
+                .build()
+                );
+    }
+
+    private final Object localTaskNoticeHelper = new Object();
+
+    @SuppressFBWarnings("NN_NAKED_NOTIFY")
+    private void noticeEnqueue()
+    {
+        synchronized (localTaskNoticeHelper) {
+            localTaskNoticeHelper.notifyAll();
+        }
+    }
+
+    private void sleepForEnqueue(long maxSleepMillis)
+    {
+        synchronized (localTaskNoticeHelper) {
+            try {
+                localTaskNoticeHelper.wait(maxSleepMillis);
+            }
+            catch (InterruptedException ex) {
+                return;
```

Comment:
Why don't you call `Thread.currentThread().interrupt()` to restore the interrupted status?